### PR TITLE
Add channels_last3d support for mkldnn conv and mkldnn deconv

### PR DESCRIPTION
**Pull request by CaoE**
### Motivation

- Add channels_last3d support for mkldnn conv and mkldnn deconv.
- Use `ideep::convolution_transpose_forward::compute_v3` instead of `ideep::convolution_transpose_forward::compute`.  compute_v3 uses `is_channels_last` to notify ideep whether to go CL or not to align with the memory format check of PyTorch.

### Testing
1 socket (28 cores):

- memory format: torch.contiguous_format

module | shape | forward / ms | backward / ms
-- | -- | -- | --
conv3d | input size: (32, 32, 10, 100, 100), weight size: (32, 32, 3, 3, 3) | 64.56885 | 150.1796
conv3d | input size: (32, 16, 10, 200, 200), weight size: (16, 16, 3, 3, 3) | 100.6754 | 231.8883
conv3d | input size: (16, 4, 5, 300, 300), weight size: (4, 4, 3, 3, 3) | 19.31751 | 68.31131

module | shape | forward / ms | backward / ms
-- | -- | -- | --
ConvTranspose3d | input size: (32, 32, 10, 100, 100), weight size: (32, 32, 3, 3, 3) | 122.7646 | 207.5125
ConvTranspose3d | input size: (32, 16, 10, 200, 200), weight size: (16, 16, 3, 3, 3) | 202.4542 | 368.5492
ConvTranspose3d | input size: (16, 4, 5, 300, 300), weight size: (4, 4, 3, 3, 3) | 122.959 | 84.62577

- memory format: torch.channels_last_3d

module | shape | forward / ms | backward / ms
-- | -- | -- | --
conv3d | input size: (32, 32, 10, 100, 100), weight size: (32, 32, 3, 3, 3) | 40.06993 | 114.317
conv3d | input size: (32, 16, 10, 200, 200), weight size: (16, 16, 3, 3, 3 | 49.08249 | 133.4079
conv3d | input size: (16, 4, 5, 300, 300), weight size: (4, 4, 3, 3, 3) | 5.873911 | 17.58647

module | shape | forward / ms | backward / ms
-- | -- | -- | --
ConvTranspose3d | input size: (32, 32, 10, 100, 100), weight size: (32, 32, 3, 3, 3) | 88.4246 | 208.2269
ConvTranspose3d | input size: (32, 16, 10, 200, 200), weight size: (16, 16, 3, 3, 3 | 140.0725 | 270.4172
ConvTranspose3d | input size: (16, 4, 5, 300, 300), weight size: (4, 4, 3, 3, 3) | 23.0223 | 37.16972

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### CODE REVIEW

**Comment by CaoE** [comment]
It depends on https://github.com/pytorch/pytorch/pull/97430

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -503,9 +503,6 @@ struct ConvParams {
     if (transposed && is_output_padding_big()) {
       return false;
     }
-    if (transposed && groups > 1 && at::symint::size<T>(input, 1) == groups) {
-      return false;
-    }
```
Comment (jgong5) [inline aten/src/ATen/native/Convolution.cpp:508]: May I know why this line is removed? It is not related to channels 3d, right?

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 2]
```diff
@@ -927,17 +928,19 @@ Tensor mkldnn_convolution_transpose(
...
+  bool use_channels_last = mkldnn_conv_use_channels_last(input_t, weight_t);
+  auto memory_format = mkldnn_convolution_memory_format(input_t.ndimension(), use_channels_last);
 
-  auto output_sizes = conv_input_size(input.sizes(), weight.sizes(), padding, output_padding, stride, dilation, groups);
-  auto output = at::empty({0}, input.options());
+  auto output_sizes = conv_input_size(input_t.sizes(), weight_t.sizes(), padding, output_padding, stride, dilation, groups);
+  auto output = at::empty({0}, input_t.options());
 
+  auto input = input_t.is_mkldnn() ? input_t : input_t.contiguous(memory_format);
+  auto weight = weight_t.is_mkldnn() ? weight_t : weight_t.contiguous(memory_format);
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp:943]: If input is already in channels last format while weight is not, is it better to not do channels last conversion here and fuse this with weight packing before invoking mkldnn conv primitive?

File: third_party/ideep
Comment (jgong5) [inline third_party/ideep:1]: Suggest to separate ideep changes

File: aten/src/ATen/native/mkldnn/Conv.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/mkldnn/Conv.cpp:943]: Add contiguous for deconv here to align with the behavior of conv (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mkldnn/Conv.cpp#L223-L224). Actually, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L1583 and https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L2169 already do contiguous for input and weight.

File: third_party/ideep
Comment (CaoE) [reply third_party/ideep:1]: Rebased, and https://github.com/pytorch/pytorch/pull/97430 is merged.  ideep changes are removed.

**Review by jgong5 — APPROVED** [review]

**Review by mingfeima — APPROVED** [review]
LGTM

**Comment by CaoE** [comment]
@malfet Could you please review this PR ? Thank you.

**Comment by mingfeima** [comment]
@CaoE https://github.com/pytorch/pytorch/pull/99539 this one will be landed first, rebase again when it is settled. and add more unit test cases for both inference and training.

**Review by mingfeima — CHANGES_REQUESTED** [review]
@CaoE please include the shape checks for illegal inputs, https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mkldnn/Conv.cpp#L91 to mkldnn transpose conv.

So that when an illegal input is given, the error will be reported from pytorch, not ondenn.

**Comment by CaoE** [comment]
> @CaoE please include the shape checks for illegal inputs, https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mkldnn/Conv.cpp#L91 to mkldnn transpose conv.


mkldnn transpose conv is not exported to python like mkldnn conv. Users can't use torch.mkldnn_convolution_transpose directly. These is already a shape check for illegal inputs in convolution.cpp for deconv. Do we still need to add shape checks for illegal inputs ?

**Comment by CaoE** [comment]
Rebased since https://github.com/pytorch/pytorch/pull/99539 merged.

**Comment by CaoE** [comment]
@cpuhrsch Could you please help to review this PR. Thank you.

**Comment by CaoE** [comment]
@peterbell10 Could you please help to review this PR. Thank you.

**Comment by CaoE** [comment]
@Skylion007 Could you please help to review this PR. Thank you.

File: test/test_mkldnn.py [hunk 3]
```diff
@@ -315,46 +318,69 @@ def _test_conv2d_nhwc_base(self, conv_module, weight_memory_format, dtype):
...
-            self._test_conv2d_nhwc_base(torch.nn.Conv2d, torch.contiguous_format, dtype=torch.bfloat16)
-            self._test_conv2d_nhwc_base(torch.nn.Conv2d, torch.channels_last, dtype=torch.bfloat16)
-
-    def test_conv_transpose2d_nhwc(self):
-        self._test_conv2d_nhwc_base(torch.nn.ConvTranspose2d, torch.contiguous_format, dtype=torch.float32)
-        self._test_conv2d_nhwc_base(torch.nn.ConvTranspose2d, torch.channels_last, dtype=torch.float32)
+            self._test_conv_deconv_nhwc_base(torch.nn.Conv2d, torch.contiguous_format, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.Conv2d, torch.channels_last, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.Conv3d, torch.contiguous_format, dtype=torch.bfloat16)
+            # skip the test until onednn updates to 3.1 because onednn will get wrong result
```
Comment (cpuhrsch) [inline test/test_mkldnn.py:354]: Seems like this is a blocking issue? We can't ship this if it isn't producing correct results. Is there a PR that attempts to update onednn in PyTorch?

Comment (CaoE) [reply test/test_mkldnn.py:354]: https://github.com/pytorch/pytorch/pull/97957 attempts to update onednn in PyTorch

Comment (cpuhrsch) [reply test/test_mkldnn.py:354]: Ok, let's wait for https://github.com/pytorch/pytorch/pull/97957 first then.

**Comment by CaoE** [comment]
@cpuhrsch https://github.com/pytorch/pytorch/pull/97957 is merged. Could you please review this PR ? Thank you.

File: test/test_mkldnn.py (hunk 3 above)
Comment (CaoE) [reply test/test_mkldnn.py]: un-skip the test since https://github.com/pytorch/pytorch/pull/97957 is merged.

**Comment by CaoE** [comment]
@atalman Could you please review This PR ? Thank you.

**Comment by CaoE** [comment]
@mikaylagawarecki Could you please review this PR ? We would suggest it to be in PT 2.1 if possible.

**Comment by atalman** [comment]
cc @albanD

File: test/test_mkldnn.py [hunk 4]
```diff
@@ -347,46 +350,67 @@ def _test_conv2d_nhwc_base(self, conv_module, weight_memory_format, dtype):
...
-    def test_conv_transpose2d_nhwc(self):
-        self._test_conv2d_nhwc_base(torch.nn.ConvTranspose2d, torch.contiguous_format, dtype=torch.float32)
-        self._test_conv2d_nhwc_base(torch.nn.ConvTranspose2d, torch.channels_last, dtype=torch.float32)
+            self._test_conv_deconv_nhwc_base(torch.nn.Conv2d, torch.contiguous_format, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.Conv2d, torch.channels_last, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.Conv3d, torch.contiguous_format, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.Conv3d, torch.channels_last_3d, dtype=torch.bfloat16)
+        # test fall back to thnn impl
+        with torch.backends.mkldnn.flags(enabled=False):
+            self._test_conv_deconv_nhwc_base(torch.nn.Conv2d, torch.contiguous_format, dtype=torch.bfloat16, prec=1e-2)
```
Comment (cpuhrsch) [inline test/test_mkldnn.py:389]: Those are some pretty high tolerances. For reference the [default value for bfloat16 is 0.016 for rtol and 1e-5 for atol](https://github.com/pytorch/pytorch/blob/main/torch/testing/_comparison.py#L62-L70).

If it's testing the fallback that already exists, why are these adjustments needed?

File: test/test_mkldnn.py [hunk 5]
```diff
@@ -347,46 +350,67 @@ def _test_conv2d_nhwc_base(self, conv_module, weight_memory_format, dtype):
...
         # when torch.ops.mkldnn._is_mkldnn_bf16_supported() returns false, bf16 CPU conv will fall back to thnn impl
         if torch.ops.mkldnn._is_mkldnn_bf16_supported():
-            self._test_conv2d_nhwc_base(torch.nn.ConvTranspose2d, torch.contiguous_format, dtype=torch.bfloat16)
-            self._test_conv2d_nhwc_base(torch.nn.ConvTranspose2d, torch.channels_last, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.ConvTranspose2d, torch.contiguous_format, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.ConvTranspose2d, torch.channels_last, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.ConvTranspose3d, torch.contiguous_format, dtype=torch.bfloat16)
+            self._test_conv_deconv_nhwc_base(torch.nn.ConvTranspose3d, torch.channels_last_3d, dtype=torch.bfloat16)
+        # test fall back to thnn impl
+        with torch.backends.mkldnn.flags(enabled=False):
```
Comment (cpuhrsch) [inline test/test_mkldnn.py:409]: Same here

**Comment by cpuhrsch** [comment]
Structurally the code looks good, just one small comment around tolerances.

File: test/test_mkldnn.py (hunk 4 above)
Comment (CaoE) [reply test/test_mkldnn.py:389]: BF16 fallback testing is not covered before. BF16 fallback implementations require higher tolerances compared to onednn's conv, as fallback implementations are divided into two parts im2col+gemm, and the number of data type conversions in the middle is more than that of onednn's direct conv, resulting in additional accuracy loss.

File: test/test_mkldnn.py (hunk 5 above)
Comment (CaoE) [reply test/test_mkldnn.py:409]: Same as above

**Comment by CaoE** [comment]
@cpuhrsch Could you please review this PR again ? Thank you.

**Review by cpuhrsch — APPROVED** [review]
Great!

**Comment by cpuhrsch** [comment]
@CaoE - You might need to rebase this PR due to unrelated failures.
